### PR TITLE
Replace or skip PCC in mean, softmax, and scalar eltwise tests

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
### PR Category
Test Only

### Issue
https://github.com/tenstorrent/tt-metal/issues/47585

## Summary

This PR splits out **test-only assertion updates** from [#42848](https://github.com/tenstorrent/tt-metal/pull/42848). These changes do **not** modify `comp_pcc`. They can land independently to keep #42848 focused on the core `comp_pcc` change and its directly coupled regressions.

The theme: use **checks that match how each op’s output is structured** — per-element tolerance (allclose), global error (Frobenius), and ULP where values are independently computed — instead of Pearson correlation where it adds little or is ill-defined (e.g. constant / single-element tensors).

### `tests/ttnn/unit_tests/operations/reduce/test_reduction_mean.py`
**Reason:** Disable PCC on constant-output mean scaling tests.

- **`test_mean_scaling`** and **`test_mean_scaling_factor`** use `torch.ones` input, so the golden output is **uniform** (same value in every element after mean × scalar).
- These tests already validate with **allclose**, **Frobenius**, and **ULP** (`check_ulp=True`).
- Mean along a dimension produces **independent per-slice values**; element-wise tolerance is the right check. A global correlation metric is redundant here and breaks down when variance is zero (every element identical).
- **Change:** `check_pcc=False` on both tests. **No new checks added** — existing allclose + Frobenius + ULP remain.

---

### `tests/ttnn/unit_tests/operations/fused/test_softmax.py`
**Reason:** Skip PCC where softmax output is degenerate (constant or single-element slice).

- **`test_softmax`:** When the reduced dimension has size 1, output is trivially constant (~1.0). PCC is undefined on zero-variance tensors. **Change:** `check_pcc=(dim_size > 1)` so PCC runs only when the slice has multiple softmax probabilities to compare. **allclose** (rtol=0.088, atol=0.009) and **Frobenius** (0.030) still run for all cases. **ULP not added** — non-constant paths still rely on the existing loose allclose/Frobenius (+ PCC when dim > 1).
- **`test_large_fill_softmax`:** Uniform fill input → **constant** softmax output (1/n). This is a numerical-stability / overflow stress test , not a correlation test. **Change:** `check_pcc=False`. **allclose** (rtol=0.008, atol=0.002) and **Frobenius** (0.008) are sufficient. **ULP not added** — tight allclose already covers the small uniform error (~1 ULP in practice).

---

### `tests/ttnn/unit_tests/operations/eltwise/test_binary_ng_typecast.py`
**Reason:** Replace PCC with checks suited to scalar eltwise ops (including non-finite results).

- **`test_edgecase_dims_eltwise_scalar_matrix_math`** previously used `assert_with_pcc` for all scalar ops.
- **Divide by scalar `0.0`:** Golden and device can disagree on **inf vs NaN** placement; PCC masks or mishandles non-finites. **Change:** compare **non-finite position masks** with `assert_close`, then **ULP on finite elements only** (`ulp_threshold=3`).
- **All other scalar ops in this test:** **Change:** `assert_with_pcc` → `assert_with_ulp(ulp_threshold=3)` for direct per-element bfloat16 accuracy.
- **Does not depend on #42848** — this is a standalone assertion improvement that works on current `main`.

Pipeline check

- [x] Sanity test
- [x] (Runtime) Sanity Tests
- [x] BHPC

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:virdhatchani/pcc_1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:virdhatchani/pcc_1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:virdhatchani/pcc_1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:virdhatchani/pcc_1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:virdhatchani/pcc_1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:virdhatchani/pcc_1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:virdhatchani/pcc_1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:virdhatchani/pcc_1)

### CODE REVIEW

**Review by Aswinmcw — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/reduce/test_reduction_mean.py [hunk 1]
```diff
@@ -66,11 +66,11 @@ def test_mean_scaling(device, shape, dim, keepdim):
     assert_numeric_metrics(
         torch_output_tensor,
         output_tensor,
-        pcc_threshold=0.999,
         rtol=0.004,
         atol=0.004,
         frobenius_threshold=0.004,
         check_ulp=True,
+        check_pcc=False,
     )
```
Comment (nmauriceTT) [inline tests/ttnn/unit_tests/operations/reduce/test_reduction_mean.py]: Here, we have relatively short tensors which are set to 1. Accumulation errors should be non-existent. 
I think we should use assert_with_ulp here. Ideally with `ulp_threshold=0` or 1.

File: tests/ttnn/unit_tests/operations/reduce/test_reduction_mean.py [hunk 2]
```diff
@@ -93,11 +93,11 @@ def test_mean_scaling_factor(device, shape, dim, scalar):
     assert_numeric_metrics(
         torch_output_tensor,
         output_tensor,
-        pcc_threshold=0.9999,
         rtol=0.004,
         atol=0.008,
         frobenius_threshold=0.004,
         check_ulp=True,
+        check_pcc=False,
     )
```
Comment (nmauriceTT) [inline tests/ttnn/unit_tests/operations/reduce/test_reduction_mean.py]: Same comment about assert_with_ulp.

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_ng_typecast.py [hunk 3]
```diff
@@ -821,7 +821,26 @@ def test_edgecase_dims_eltwise_scalar_matrix_math(input_shape, scalar, ttnn_fn,
...
+        torch.testing.assert_close(
+            g_nonfinite, d_nonfinite, msg="Non-finite positions differ between golden and device"
+        )
+        finite_mask = torch.isfinite(torch_output_tensor) & torch.isfinite(tt_output_tensor)
+        if finite_mask.any():
+            assert_with_ulp(
+                torch_output_tensor[finite_mask],
+                tt_output_tensor[finite_mask],
+                ulp_threshold=3,
+            )
```
Comment (nmauriceTT) [inline tests/ttnn/unit_tests/operations/eltwise/test_binary_ng_typecast.py]: If `scalar` == 0 then all output values should be non-finite (-inf, +inf or NaN). 

So, unless if I am wrong, `finite_mask` should be empty. If that's the case then you can remove this section.

**Review by nmauriceTT — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]
